collagraph/collagraph.py

```diff
@@ -204,35 +225,52 @@ def update_class_component(self, fiber: Fiber):
 
         # Attach the component instance to the fiber
         fiber.component = component
-        fiber.component_watcher = watch(
-            lambda: fiber.component.state,
-            lambda: self.state_updated(fiber),
-            deep=True,
-            sync=self.event_loop_type is EventLoopType.SYNC,
-        )
+
+        component._slots = fiber.children if isinstance(fiber.children, dict) else {}
 
         if fiber.alternate:
             fiber.alternate.component = None
-            fiber.alternate.component_watcher = None
-
-        component._slots = fiber.children if isinstance(fiber.children, dict) else {}
+            fiber.watcher = fiber.alternate.watcher
+            fiber.alternate.watcher = None
+            # Re-evaluate the watcher to get the new value
+            fiber.watcher.evaluate()
+        else:
+            fiber.watcher = watch(
+                component.render,
+                lambda: self.state_updated(fiber),
+            )
 
-        # List of VNodes
-        children = [component.render()]
-        self.reconcile_children(fiber, children)
+        self.reconcile_children(fiber, [fiber.watcher.value])
 
     def update_function_component(self, fiber: Fiber):
-        if isinstance(fiber.children, dict):
-            children = [fiber.type(fiber.props, fiber.children)]
-        else:
-            children = [fiber.type(fiber.props)]
-        self.reconcile_children(fiber, children)
+        if fiber.alternate:
+            fiber.alternate.watcher = None
+
+        def render():
+            if isinstance(fiber.children, dict):
+                return [fiber.type(fiber.props, fiber.children)]
+            return [fiber.type(fiber.props)]
```

Comment:
do you have a test to see if this render function re-evaluates when the isinstance check toggles?

Reply:
No, I don't have a test for it. This is a bit of a special case: the only time that `fiber.children` is a dict, is when the children are provided as slot content. You can't specify children for a Component in any other way, so it is not possible to change the type.

Reply:
I see.